File: Difficulty: Medium/Distance of nearest cell having 1/distance-of-nearest-cell-having-1.py
from collections import deque
class Solution:
    #Function to find distance of nearest 1 in the grid for each cell.
	def nearest(self, grid):
		#Code here
		n = len(grid)
		m=len(grid[0])
		dis= [[0]*m for _ in range(n)]
		vis= [[0]*m for _ in range(n)]
		q= deque()
		
		for i in range(n):
		    for j in range(m):
		        if grid[i][j]==1:
		            q.append((i,j,0))
		            vis[i][j]=1
		delr = [-1,0,1,0]
		delc = [0,1,0,-1]
		while(q):
		    r,c,dist=q.popleft()
		    dis[r][c]=dist
		    for i in range(4):
		        nr = r+delr[i]
		        nc = c+delc[i]
		        
		        if(0<=nr<n and 0<=nc<m and not vis[nr][nc] and grid[nr][nc]!=1):
		            q.append((nr,nc,dist+1))
		            vis[nr][nc]=1
		return dis

File: Difficulty: Medium/Distance of nearest cell having 1/distance-of-nearest-cell-having-1.cpp
class Solution {
  public:
    // Function to find distance of nearest 1 in the grid for each cell.
    vector<vector<int>> nearest(vector<vector<int>>& grid) {
        // Code here
        int n = grid.size();
        int m = grid[0].size();
        
        vector<vector<int>> vis(n,vector<int>(m,0));
        vector<vector<int>> dist(n,vector<int>(m,0));
        queue<pair<pair<int,int>,int>> q;
        
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==1){
                    q.push({{i,j},0});
                    vis[i][j]=1;
                }
                else{
                    vis[i][j]=0;
                }
            }
        }
        int delr[] = {-1,0,+1,0};
        int delc[] = {0,+1,0,-1};
        while(!q.empty()){
            int r = q.front().first.first;
            int c = q.front().first.second;
            int steps = q.front().second;
            q.pop();
            dist[r][c]=steps;
            for(int i=0;i<4;i++){
                int nr = r+delr[i];
                int nc = c+delc[i];
                if(nr<n && nr>=0 && nc>=0 && nc<m && vis[nr][nc]==0){
                    vis[nr][nc]=1;
                    q.push({{nr,nc},steps+1});
                }
            }
        }
        return dist;
    }
};